### DISCUSSION TITLE
Question on declaring vector of matprops

### DISCUSSION THREAD

**Discussion by TheGreatCid** [idaholab/moose#28918] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hi all,

I just want a sanity check regarding the declaration of mat props in a vector. I believe I am doing things correctly, however I am seeing a seg fault during material property declaration. 

Here is a description of my method,

For the class def:
```
#include "Material.h"
#include "BaseNameInterface.h"
#include "MaterialProperty.h"
#include "RankTwoTensorForward.h"

class TestClass : public Material, public BaseNameInterface
public:
  static InputParameters validParams();
  TestClass(const InputParameters & parameters);
protected:
  std::vector<ADMaterialProperty<RankTwoTensor> *> _matvec;
```
and constructor:
```
TestClass::TestClass(const InputParameters & parameters):
    Material(parameters),
    BaseNameInterface(parameters)
{
  for (unsigned int i = 0; i <= 5; i++)
        _matvec[i] = &Material::declareADProperty<RankTwoTensor>("test" + std::to_string(i));
}

```

This is just a paired-down example, but I want to double-check that this is the correct way to do this.

Thank you!

**Comment by GiudGiud** [comment]
Hello

You need to resize the vector before accessing it at a component

Vectors are by default initialized at size 0

**Reply by TheGreatCid** [reply]
I actually did do that; I just failed to include it in this mock-up example. My bad.

So what I actually have is 
```
TestClass::TestClass(const InputParameters & parameters):
    Material(parameters),
    BaseNameInterface(parameters)
{
 _matvec.size(5);
  for (unsigned int i = 0; i <= 5; i++)
        _matvec[i] = &Material::declareADProperty<RankTwoTensor>("test" + std::to_string(i));
}

```

**Reply by TheGreatCid** [reply]
Is there an obvious source of a seg fault here? If not, then it's probably something specific that I trimmed out of this example, and I will continue my search.

**Reply by GiudGiud** [reply]
.resize not .size

**Reply by GiudGiud** [reply]
also i =0 to i =5 is 6 not 5 elements

**Reply by TheGreatCid** [reply]
This example I drafted is falling apart. The issues you pointed out are not present in the actual code(of course, you could not have known that). Out-of-index issues aside, is this the correct way to initialize a vector of mat props?

**Reply by GiudGiud** [reply]
It seems fine. 

Do you fill in these tensors later on? Where is the seg fault? Do you have a backtrace?

**Reply by TheGreatCid** [reply, marked as answer]
I figured it out. It was unrelated to how the vector was set up. The value I was using to set the size of the vec was being set to zero and I had not noticed that.
